### PR TITLE
feat: add copilot impl

### PR DESCRIPTION
**Pull request by darkskygit**
fix CLOUD-22
fix CLOUD-24

### CODE REVIEW

**Comment by darkskygit** [comment]
* **#6528** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6528?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6527** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6527?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6506** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6506?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6525** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6525?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6445** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6445?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6448** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6448?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6437** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6437?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6272** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6272?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6230** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6230?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#6254** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6254?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6241** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6241?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#6229** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6229?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @darkskygit and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by zkwolf** [comment]
I think it's possible to have a less strict restriction on model names, so that in the future users can use some openai api-compatible services, such as ollama and groq.

**Comment by darkskygit** [comment]
> I think it's possible to have a less strict restriction on model names, so that in the future users can use some openai api-compatible services, such as ollama and groq.

will have another provider impl for that -- this will make the code more maintainable, although most of the logic and API are similar

File: .github/helm/affine/charts/graphql/templates/deployment.yaml [hunk 1]
```diff
@@ -83,6 +83,8 @@ spec:
             value: "{{ .Values.app.features.earlyAccessPreview }}"
           - name: FEATURES_SYNC_CLIENT_VERSION_CHECK
             value: "{{ .Values.app.features.syncClientVersionCheck }}"
+          - name: FEATURES_COPILOT_SKIP_AUTHORIZATION
```
Comment (forehalo) [inline .github/helm/affine/charts/graphql/templates/deployment.yaml]: should be moved to public endpoint pr

File: packages/backend/server/src/config/affine.env.ts [hunk 2]
```diff
@@ -37,5 +37,9 @@ AFFiNE.ENV_MAP = {
     'featureFlags.syncClientVersionCheck',
     'boolean',
   ],
+  FEATURES_COPILOT_SKIP_AUTHORIZATION: [
```
Comment (forehalo) [inline packages/backend/server/src/config/affine.env.ts]: move

File: packages/backend/server/src/plugins/copilot/providers/openai.ts [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+  ChatMessage,
+  ChatMessageRole,
+  CopilotCapability,
+  CopilotProviderType,
+  CopilotTextToEmbeddingProvider,
+  CopilotTextToTextProvider,
+} from '../types';
+
+export class OpenAIProvider
+  extends OpenAI
```
Comment (forehalo) [inline packages/backend/server/src/plugins/copilot/providers/openai.ts]: The provider is limited by interfaces, nonsense to extends OpenAI, make it a field instead

**Review by forehalo — APPROVED** [review]

**Pull request merged**